Add support for Fly.io deployments (#167)
elixir-webrtc/ex_webrtc: guides/deploying/fly.md

# Deploying on Fly.io

Elixir WebRTC-based apps can be easily deployed on [Fly.io](https://fly.io)!

There are just two things you need to do:

- configure a STUN server both on the client and server side
- use custom Fly.io IP filter on the server side

In theory, configuring a STUN server just on a one side should be enough but we recommend to do it on both sides.

In JavaScript code:

```js
pc = new RTCPeerConnection({
  iceServers: [{ urls: "stun:stun.l.google.com:19302" }],
});
```

> in Elixir code:

Reviewer: ```suggestion
In Elixir code:
```

```elixir
ip_filter = Application.get_env(:your_app, :ice_ip_filter)

{:ok, pc} =
  PeerConnection.start_link(
    ice_ip_filter: ip_filter,
    ice_servers: [%{urls: "stun:stun.l.google.com:19302"}]
  )

> ```

Reviewer: [nitpick] When `:ice_ip_filter` is unset in the application, we're passing `ice_ip_filter: nil` to `PeerConnection`, which means https://github.com/elixir-webrtc/ex_webrtc/blob/e42c09df1ae702126dc407ee35a5760d0f601967/lib/ex_webrtc/peer_connection/configuration.ex#L242 

doesn't insert the default filter.

When we pass it to ICE, however, this line
https://github.com/elixir-webrtc/ex_ice/blob/4c5190687e11b28a8209f2c930895647d0b2a64c/lib/ex_ice/priv/ice_agent.ex#L137
makes it work again, as it gets the `nil` stored in the keyword list and overrides it with the default filter.

I feel like this creates a kind of ambiguity and depends on ExICE not changing its behaviour of parsing options, which can potentially lead to hard-to-find errors in the future. To avoid this in this example, I would explicitly *not* insert `ice_ip_filter` to the PC options when it's missing from the application's environment.

Author: Good catch!

> in `runtime.exs`:

Reviewer: ```suggestion
In `runtime.exs`:
```

```elixir
if System.get_env("FLY_IO") do
  config :your_app, ice_ip_filter: &ExWebRTC.ICE.FlyIpFilter.ip_filter/1
end
```

> in fly.toml:

Reviewer: ```suggestion
In fly.toml:
```

```toml
[env]
  # add one additional env
  FLY_IO = 'true'
```

That's it!

> No special UDP port exports or dedicated IP address needed.

Reviewer: ```suggestion
No special UDP port exports or dedicated IP addresses are needed.
```
